controlplane/kubeadm/internal/proxy/dial.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+const defaultTimeout = 10 * time.Second
+
+// Dialer creates connections using Kubernetes API Server port-forwarding
+type Dialer struct {
+	proxy          Proxy
+	clientset      *kubernetes.Clientset
+	proxyTransport http.RoundTripper
+	upgrader       spdy.Upgrader
+	timeout        time.Duration
+	requestIDLock  sync.Mutex
+	requestID      int
+}
+
+// NewDialer creates a new dialer for a given API server scope
+func NewDialer(p Proxy, options ...func(*Dialer) error) (*Dialer, error) {
+	if p.Port == 0 {
+		return nil, errors.New("port required")
+	}
+
+	dialer := &Dialer{
+		proxy: p,
+	}
+
+	for _, option := range options {
+		err := option(dialer)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if dialer.timeout == 0 {
+		dialer.timeout = defaultTimeout
+	}
+	p.KubeConfig.Timeout = dialer.timeout
+	clientset, err := kubernetes.NewForConfig(p.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	proxyTransport, upgrader, err := spdy.RoundTripperFor(p.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	dialer.proxyTransport = proxyTransport
+	dialer.upgrader = upgrader
+	dialer.clientset = clientset
+	return dialer, nil
+}
+
+// DialContextWithAddr is a GO grpc compliant dialer construct
+func (d *Dialer) DialContextWithAddr(ctx context.Context, addr string) (net.Conn, error) {
+	return d.DialContext(ctx, scheme, addr)
+}
+
+// nextRequestID generates the next request ID for calling to the API server
+func (d *Dialer) nextRequestID() int {
```

Comment:
The request ID is scoped to a single upgraded httpstream/spdy connection, and ideally not the dialer. The reason the request ID exists is that if you're port forwarding and a client makes several connections in rapid succession (such as a web browser using multiple connections to retrieve html + css + images + js), the streaming code needs to be able to associate the (data stream, error stream) pair for each forwarded connection (all going across a single upgraded spdy connection).

I assume we can probably use a hard-coded request id header of 0 or 1 for each call to DialContext. WDYT?

Reply:
0 makes sense.